added options in cfg file to show or hide remote information

casonadams opened the pull request.

arl commented on format/tmux/formater.go:
```diff
@@ -17,7 +17,8 @@ type Config struct {
 	// components.
-	Styles styles
+	Styles  styles
+	Display display
```
Can you add a godoc-friendly comment here.
Please see the differences between:

    go doc ./format/tmux Config.Symbols

and

    go doc ./format/tmux Config.Display

arl commented on format/tmux/formater.go:
```diff
@@ -33,6 +34,7 @@ type symbols struct {
 	Untracked string // Untracked is the string shown before the count of untracked files.
 	Stashed   string // Stashed is the string shown before the count of stash entries.
 	Clean     string // Clean is the string shown when the working tree is clean.
+	Delimiter string // Delimiter is the string shown before flags
```
Since the order of components can now be re-ordered, I guess the Delimiter should represent the string that delimits them.

arl commented on format/tmux/formater.go:
```diff
@@ -47,6 +49,10 @@ type styles struct {
 }
 
+type display struct {
+	Output []string // Display order and visiblity of options
```
Please make this comment godoc-friendly

arl commented on format/tmux/formater.go:
```diff
@@ -88,15 +98,20 @@ func (f *Formater) Format(w io.Writer, st *gitstatus.Status) error {
 		fmt.Fprintf(w, "%s%s [no commits yet]", f.Styles.Branch, f.st.LocalBranch)
-		goto fileCounts
+		f.flags()
+	} else {
```
Instead of adding else, we could write the flags, write the state (WriteTo) and return.
Doing so you'd both have simplified your code and mine (removing `goto`) ! :)

arl commented on format/tmux/formater.go:
```diff
@@ -164,7 +179,7 @@ func (f *Formater) divergence() {
 func (f *Formater) flags() {
 	f.clear()
-	f.b.WriteString(" - ")
+	f.b.WriteString(f.Symbols.Delimiter)
```
We can't just write the delimiter before the flags, since the flags could be the first component to be shown if the user has placed the `flags` component first in their configuration file. It will need to be intertwined in-between both components.

arl commented on format/tmux/formater_test.go:
```diff
@@ -58,6 +60,7 @@ func TestFormater_flags(t *testing.T) {
 			symbols: symbols{
 				Conflict:  "SymbolConflict",
 				Untracked: "SymbolUntracked",
+				Delimiter: " - ",
```
I'd like to see at least a test case with another delimiter than the default one, to check we're actually showing the strings from the config.
Plus, as said in the issue, I'd like to see at least a test where we expect the components to be shown in reverse order with respect to the default

casonadams replied:
The tests are more of an integration level test.  They may be some what fragile due to the nature of git branch naming.  I believe I have covered most naming conventions.  Also had to get a little creative to get travis CI to work.  It doesn't ever get remote branch data.  Let me know what you think.

arl commented:
I don't have the time to do a full review at the moment, sorry.
First of all, thank you for contributing, this is very much appreciated!

About your recent set of changes, I see that you have added a second delimiter. I'm sorry I'd prefer to keep this change more simple than that, please re-read my first review, we can always build up on this afterwards.

> To keep it simple, I'd prefer to just have 2 possible components in the Display struct: `branch` and `flags` where `branch` would be made of both the local and the origin branch

Please tell me something was not clear in my previous request.

I'm taking precautions because, even if `gitmux` enjoys a very small community of users, they still exist, so I prefer to introduce with caution any change in the configuration file format with, I'd not want future developments/versions to be impacted by previous design choices.

About your question in the test: look at other tests in the same file: you just have to fill a `gitstatus.Status` structure with the pieces you need, create a tmux formater with it, call the appropriate method that modifies the internal state of `tmux.Formater`, then compare the expected output with `f.b.String()`.

casonadams commented:
There was nothing unclear in your request.  Sorry that I am not building the functions that you desire and or the community around this project.  

My goal was to remove the remote branch info.  As I don't like the amount of space it is taking up in my status line.  

I do understand changes like these can break others that are using customized cfg files.

With that in mind, let me know if you want me to continue to pursue some changes like this or if I should close this PR for now.

Thanks and sorry for the miss communications.

arl commented:
> There was nothing unclear in your request. Sorry that I am not building the functions that you desire and or the community around this project.
> 
> My goal was to remove the remote branch info. As I don't like the amount of space it is taking up in my status line.

I understand now that you only wanted to hide remote information. That's fair, it's a good feature I'm sure you're not the only one to desire. I'm also often finding it to eat up too much space with projects with long branch names.

So this is where the misunderstanding came from: the issue I had in mind was more about giving full control over ordering and presence of each information shown in the status line. I also thought of automatically removing the remote when there's not enough space to show it. Anyway, this requires a bit of work and was not the object of the PR.
 
In that case, please let me propose a simple way to hide the remote: why not keeping the new `Display` section, but let's forget about the `Display.Order` field, there would just be a single boolean in that section, `Display.Remote`, which would indicate, when true, that the remote should be shown.

> I do understand changes like these can break others that are using customized cfg files.
> 
> With that in mind, let me know if you want me to continue to pursue some changes like this or if I should close this PR for now.
> 
> Thanks and sorry for the miss communications.
 
Please tell me what you think of the above proposition. I'd be happy to review a PR of yours implementing it. If you wish to do so, please go ahead, either on this PR or another one. Or just close it if you're not.
Thank you for the time you're involving in this project.
Sorry for the misunderstanding.

casonadams commented:
Okay so I gave a bit more thought on how to make the old config work with the new features.

My thoughts are in the `display: []` section any order of branch, remote, or flags including custom dividers.  This really allows for a lot of flexibility without having breaking changes from previous configs.  And if the display section has a typo it will just 

Let me know your thoughts.  I would love to figure out how to get a feature like this into this project.

`example config`
```
tmux:
  symbols:
    branch: ' '
    hashprefix: ':'
    ahead: ↑·
    behind: ↓·
    staged: ●
    conflict: '✖ '
    modified: '✚ '
    untracked: …
    stashed: '⚑ '
    clean: ✔
    delimiter: ' '
  styles:
    state: '#[fg=red,nobold]'
    branch: '#[fg=white,nobold]'
    remote: '#[fg=cyan]'
    staged: '#[fg=green,nobold]'
    conflict: '#[fg=red,nobold]'
    modified: '#[fg=red,nobold]'
    untracked: '#[fg=magenta,nobold]'
    stashed: '#[fg=cyan,nobold]'
    clean: '#[fg=green,nobold]'
  display: [branch, '..', remote, ' - ', flags]
```

arl commented on README.md:
```diff
@@ -67,9 +67,10 @@ Modify the line in `.tmux.conf`, passing the path of the configuration file as a
+`gitmux` configuration is split into 3 sections:
  - symbols: they are just strings of unicode characters
  - styles: they are tmux format strings (`man tmux` for reference)
+ - display: defines the layout in the form of a YAML array, seperators should be surrounded in single quotes example: `display: [branch, '..', remote, ' - ', flags]`
```
Typo seperator -> separator

arl commented on format/tmux/formater.go:
```diff
@@ -18,6 +18,8 @@ type Config struct {
 	// components.
 	Styles styles
+	// Display sets the output format of the Git status.
+	Display []string `yaml:",flow"`
```
That's a great idea

arl commented on format/tmux/formater_test.go:
```diff
@@ -156,3 +161,148 @@ func TestFormater_divergence(t *testing.T) {
+				Config: Config{Styles: tc.styles, Symbols: tc.symbols, Display: tc.display},
+				st:     tc.st,
+			}
+			st, _ := gitstatus.New()
```
Can we do like in the other tests (i.e without having to run `git` under the hood)?

arl replied:
I mean. Tests should not depend on the git status of current directory, tests should only be test formatting of a specially crafted git.Status struct with the tmux formatter

arl commented on go.sum:
```diff
@@ -4,6 +4,7 @@ github.com/davecgh/go-spew v1.1.0 h1:ZDRjVQ15GmhC3fiQ8ni8+OwkZQO4DARzQgrnXU1Liz8
 github.com/davecgh/go-spew v1.1.0/go.mod h1:J7Y8YcW2NihsgmVo/mv3lAwl/skON4iLHjSsI+c5H38=
 github.com/pmezard/go-difflib v1.0.0 h1:4DBwDE0NGyQoBHbLQYPwSUPoCMWR5BEzIk/f1lZbAQM=
 github.com/pmezard/go-difflib v1.0.0/go.mod h1:iKH77koFhYxTK1pcRnkKkqfTogsbg7gZNVY4sRDYZ/4=
+github.com/stretchr/objx v0.1.0 h1:4G4v2dO3VZwixGIRoQ5Lfboy6nUhCyYzaqnIAPPhYs4=
```
Can we have this removed?

salarkhan replied:
looks like not, objx is a dep of testify: https://github.com/stretchr/testify/blob/37e2176163624ceb4edeed061b7fc554f468f107/go.mod

```
salarkhan@saladpad:~/code/gitmux: go list -m all
github.com/arl/gitmux
github.com/arl/gitstatus v0.4.2
github.com/davecgh/go-spew v1.1.0
github.com/pmezard/go-difflib v1.0.0
github.com/stretchr/objx v0.1.0
github.com/stretchr/testify v1.3.0
gopkg.in/check.v1 v0.0.0-20161208181325-20d25e280405
gopkg.in/yaml.v2 v2.2.4
```

salarkhan commented:
@casonadams you mind if i take this over the line? i'd love for this to be on master && it looks to be 99% done already!

arl commented:
@salarkhan please do. I'd be happy to review and merge this ASAP

arl commented:
@casonadams +@salarkhan thank to both of you
